z3/include/z3_solver.h

```diff
@@ -26,55 +42,76 @@ namespace smt {
-  Term make_value(int64_t i, const Sort & sort) const override;
-  Term make_value(const std::string val,
-                  const Sort & sort,
-                  unsigned int base = 10) const override;
-  Term make_value(const Term & val, const Sort & sort) const override;
-  Term make_term(const std::string s, const Sort & sort) override;
+                 const Sort & sort1,
+                 const Sort & sort2,
+                 const Sort & sort3) const override;
+  Sort make_sort(SortKind sk, const SortVec & sorts) const override;
+  Sort make_sort(const Sort & sort_con, const SortVec & sorts) const override;
+  Sort make_sort(const DatatypeDecl & d) const override;
+
+  DatatypeDecl make_datatype_decl(const std::string & s) override;
+  DatatypeConstructorDecl make_datatype_constructor_decl(
+      const std::string s) override;
+  void add_constructor(DatatypeDecl & dt, const DatatypeConstructorDecl & con) const override;
+  void add_selector(DatatypeConstructorDecl & dt, const std::string & name, const Sort & s) const override;
+  void add_selector_self(DatatypeConstructorDecl & dt, const std::string & name) const override;
+  Term get_constructor(const Sort & s, std::string name) const override;
+  Term get_tester(const Sort & s, std::string name) const override;
+  Term get_selector(const Sort & s, std::string con, std::string name) const override;
+
+  Term make_term(bool b) const override;
+  Term make_term(int64_t i, const Sort & sort) const override;
+  Term make_term(const std::string val,
+                 const Sort & sort,
+                 uint64_t base = 10) const override;
+  Term make_term(const Term & val, const Sort & sort) const override;
+  Term make_symbol(const std::string name, const Sort & sort) override;
+  Term make_param(const std::string name, const Sort & sort) override;
   /* build a new term */
   Term make_term(Op op, const Term & t) const override;
   Term make_term(Op op, const Term & t0, const Term & t1) const override;
   Term make_term(Op op,
                  const Term & t0,
                  const Term & t1,
                  const Term & t2) const override;
-  Term make_term(Op op, const std::vector<Term> & terms) const override;
+  Term make_term(Op op, const TermVec & terms) const override;
   void reset() override;
   void reset_assertions() override;
-  bool has_symbol(const std::string name) const override;
-  Term lookup_symbol(const std::string name) const override;
-  void dump_smt2(FILE * file) const override;
-  // helpers
-  ::CVC4::api::OpTerm make_op_term(Op op) const;        // * * * * * * * * * * * * * * * * * * *
+  Term substitute(const Term term,
+                  const UnorderedTermMap & substitution_map) const override;
+  void dump_smt2(std::string filename) const override;
 
  protected:
-  ::z3::solver solver;
-  // keep track of created symbols for has_symbol and lookup_symbol
-  std::unordered_map<std::string, Term> symbols;
+  mutable z3::context ctx;
```

Comment:
This is fine for now, but I'm wondering if we should just change the interface instead of using `mutable` to avoid the consequences of `const`. That would require changing it for every solver, so don't worry about it now, but we can discuss it later (after the initial z3 implementation is merged). Which const methods were causing problems?

Reply:
Every time you create a sort, function, or expression in z3 you do so via a context. Without this being mutable you cannot do any of this functionality (calling a function on ctx) which I need to do often. Specifically all of the make_sort methods that I've implemented so far were causing problems

Reply:
I see, so if we changed the interface, then pretty much every method would have to be non-`const`. Yeah that's a pain. Leaving it `mutable` makes sense then!